### ISSUE TITLE
Random segfaults? (probably not an issue for here)

### ISSUE THREAD

**Issue by o-jasper** [luapower/thread#5]
I get (apparently)random Segfaults..(and odd "double free or corruption"), seen it for this, and [my Sql](https://github.com/o-jasper/lua_Searcher/tree/master/Searcher/Sql) aswel though... (its luajit 2.0.4-1)

Wondering if it is a luajit bug... Don't know where to submit this issue to luajit, so asking here to see if others get the bug with this, or with my lib. (the mailing list? dont like those much.. Might be that i need to change email client..)

See it on `pthread_test.lua` too. Luckily my tox bindings don't seem to have trouble with it.

**Comment by capr** [comment]
Hi Jasper, 

Just running pthread_test.lua segfaults? On what OS/arch? Can you try to run the tests inside pthread_test.lua individually to see which one crashes so we can narrow it down? Thanks.

**Comment by o-jasper** [comment]
It is on Arch Linux, luajit version 2.0.4-1..

Fine, dont see crashes: `test_priority_range`, `test_thread_self_equal`, `test_priorities`, `test_stress_test`.

`test_mutex` segfaults often, also saw a "double free" error;

```
*** Error in `luajit': double free or corruption (fasttop): 0x00007feab80008c0 ***
```

`test_cond_var` segfaults often, also saw:

```
 PANIC: unprotected error in call to Lua API (./pthread.lua:9: )
```

`test_rw_lock` segfaults, also saw "double free".

`while luajit thread_test.lua; do echo; done`(it seems random/rare) for the `_creation` test yield no problems; `test_queue(1000, 10,  1000, 10,  1000)` segfaults quite often, while  `test_queue(1000,  1, 10000,  1, 10000)`  very infrequently, similar accross the other functions, probably simply about number of times something occurs.

Note that with Sql i also tried to chase it down. Though i haven't had much tests on that. Not much wiser from it. It _might_ have to do with a pointer-of-pointer, somehow.

**Comment by capr** [comment]
Can you try using the luajit binary supplied with luapower and see if you get the same errors?

You can get it with `./mgit clone luajit`

Thanks.

**Comment by o-jasper** [comment]
Using the stuff from    https://github.com/luapower/luajit, the scripts dont seem pass on `LUA_PATH`? So went straight to the binary. Then this way, no problems!..

```
while luajit/bin/linux64/luajit-bin pthread/pthread_test.lua ; do echo; done
```

Would say "ah slightly newer version fixed bug", but the one compiled myself does crash..

```
luajit/csrc/luajit/src/src/luajit pthread/pthread_test.lua
```

The binaries differ too. I made mine with `make`.. however, doing:

```
make P=linux64 CFLAGS="-pthread" LDFLAGS="-pthread -s -static-libgcc -Wl,-rpath,'\$\$ORIGIN'" \
X0=luajit X=luajit-bin D=libluajit.so A=libluajit.a
```

It _does_ work.. it has to do with it is compiled somehow. It specifically mentions pthread? (might suggest that the Sql bug i experience might not be solved by this... I mean, not going to load everything into the luajit binary..)

**Comment by capr** [comment]
Yes, you need to compile luajit with "-pthread". There's a thread on the luajit mailing list on why this isn't the default (I forgot the reason). I added a note on the docs about that.

> not going to load everything into the luajit binary

Can you elaborate on this?

**Comment by capr** [comment]
>  the scripts dont seem pass on LUA_PATH?

The launcher scripts don't pass LUA_PATH, but change it to make the luapower distribution self-contained. This is a feature of luapower, not a bug :) Of course you are free to change the launcher scripts anyway you like, they're very simple scripts.

**Comment by o-jasper** [comment]
Good point with the feature/not bug :)

So apparently Arch linux doesnt use `-pthread` in compiling it? Their policy is to "take things as they come" as i understand it, though.. And the makefile that comes with what you luajit repo provides also doesnt do `-pthread`. So luajit in essence probably decided it? Bit confused how the mailing thread and what was distributed "came to different conclusions".

With "not going to load everything into the luajit binary" i mean, well `-pthread` compiles luajit with that. Whereas without it, the ffi somehow gets it itself from the `.so` file, maybe somehow the one triggers the bug and the other doesnt.

Dunno at this point if `-pthread` simply a convenience, or it does it include it differently? Havent tried `-lsqlite3` yet, dunno if that would work.

**Comment by capr** [comment]
About `-pthread` not being the default, don't worry about it, Mike always knew it's needed, he just didn't want it to be a dependency on it if it's not actually used. Plus it adds a negligible start-up overhead.
